client/consensus/manual-seal/src/rpc.rs

```diff
@@ -0,0 +1,212 @@
+	pub hash: Hash,
+	/// some extra details about the import operation
+	pub aux: ImportedAux
+}
+
+impl<Hash> ManualSeal<Hash> {
+	/// Create new `ManualSeal` with the given reference to the client.
+	pub fn new(import_block_channel: mpsc::Sender<EngineCommand<Hash>>) -> Self {
+		Self { import_block_channel }
+	}
+}
+
+impl<Hash: Send + 'static> ManualSealApi<Hash> for ManualSeal<Hash> {
+	fn create_block(
+		&self,
+		create_empty: bool,
+		finalize: bool,
+		parent_hash: Option<Hash>
+	) -> FutureResult<CreatedBlock<Hash>> {
+		let mut sink = self.import_block_channel.clone();
+		let future = async move {
+			let (sender, receiver) = oneshot::channel();
+			let result = sink.send(
+				EngineCommand::SealNewBlock {
+					create_empty,
+					finalize,
+					parent_hash,
+					sender: Some(sender),
+				}
+			).await;
+			result.map_err(map_error)?;
+
+			match receiver.await {
+				// all good
+				Ok(Ok(block)) => Ok(block),
+				// error from the authorship task
+				Ok(Err(e)) => {
+					Err(Error {
+						code: ErrorCode::ServerError(500),
+						message: format!("{}", e),
+						data: None,
+					})
+				}
+				// channel has been dropped
+				Err(_) => {
+					Err(Error {
+						code: ErrorCode::ServerError(500),
+						message: "Consensus process is terminating".into(),
+						data: None,
+					})
+				}
+			}
+		};
+
+		Box::new(Box::pin(future).compat())
+	}
+
+	fn finalize_block(&self, hash: Hash, justification: Option<Justification>) -> FutureResult<()> {
+		let mut sink = self.import_block_channel.clone();
+		let future = async move {
+			let (sender, receiver) = oneshot::channel();
+			let result = sink.send(
+				EngineCommand::FinalizeBlock { hash, sender: Some(sender), justification }
+			).await;
+			result.map_err(map_error)?;
+
+			match receiver.await {
+				// all good
+				Ok(Ok(())) => Ok(()),
+				// error from the authorship task
+				Ok(Err(e)) => {
+					Err(Error {
+						code: ErrorCode::ServerError(500),
+						message: format!("{}", e),
+						data: None
+					})
+				}
+				// channel has been dropped
+				Err(_) => {
+					Err(Error {
+						code: ErrorCode::ServerError(500),
+						message: "Consensus process is terminating".into(),
+						data: None
+					})
+				}
+			}
+		};
+
+		Box::new(Box::pin(future).compat())
+	}
+}
+
+/// convert a futures::mpsc::SendError to jsonrpc::Error
+fn map_error(error: SendError) -> Error {
+	if error.is_disconnected() {
+		log::warn!("Received sealing request but Manual Sealing task has been dropped");
+	}
+
+	Error {
+		code: ErrorCode::ServerError(500),
+		message: "Consensus process is terminating".into(),
+		data: None
+	}
+}
+
+/// report any errors or successes encountered by the authorship task back
+/// to the rpc
+pub fn send_result<T: std::fmt::Debug>(
+	sender: Sender<T>,
+	result: std::result::Result<T, crate::Error>
+) {
+	sender.map(|sender| {
+		sender.send(result)
+			.expect("receiving end isn't dropped until it receives a message; qed")
```

Comment:
I guess it might be when the node is terminating, best to just print an error here instead of panicking imho.